File: packages/mind/toolkit/place/scope.py
"""Scope resolvers -- each returns the `.claude/` root an artifact lands in, with
the explicit input that scope requires. The artifact is identical across scopes;
scope is the accident ([[scope-grant]]).

  user     home resolver       -> <home>/.claude        (home defaults to $HOME)
  project  project resolver    -> <project>/.claude     (project defaults to cwd)
  fleet    hosts[]             -> remote user scope per host (see place.ssh)
  local    project, unversioned-> <project>/.claude     (SEAM: gitignored; deferred)

fleet composes the per-host remote placer; local composes project scope plus a
gitignore concern not yet needed by any kind (agents/skills are versioned).
"""
from __future__ import annotations

import pathlib
import sys


def user_scope(home: str | None = None) -> pathlib.Path:
    """User scope claude dir: <home>/.claude. `--home` is the user's HOME dir, so a
    bare home self-corrects -- `.claude` appended, LOUDLY -- while a path already
    ending in `.claude` is used verbatim (back-compat). This mirrors the remote
    placer's `_resolve_claude_dir`: before, LOCAL took an explicit `--home`
    verbatim, so `--home /Users/lex` littered `/Users/lex/{agents,skills}` beside
    the real `~/.claude` -- the same footgun fixed for ssh in f6197f8, here closed
    for the local path too. No `--home` -> $HOME/.claude."""
    if not home:
        return pathlib.Path.home() / ".claude"
    p = pathlib.Path(home).expanduser()
    if p.name == ".claude":
        return p
    print(f"  NOTE --home {home!r} is a home dir -> deploying to {p / '.claude'}",
          file=sys.stderr)
    return p / ".claude"


def project_scope(project: str | None = None) -> pathlib.Path:
    """Project scope claude dir: <project>/.claude (project defaults to cwd)."""
    root = pathlib.Path(project).expanduser() if project else pathlib.Path.cwd()
    return root / ".claude"


# fleet: the priority hosts deploy ships to (SSH user differs per host -- see
# the fleet topology). A list of (user, host); the CLI loops place.ssh over them.
def local_scope(project: str | None = None) -> pathlib.Path:
    """Local (unversioned) scope -- SEAM. Same dir as project scope today; the
    distinguishing gitignore concern is deferred until a kind needs it."""
    return project_scope(project)

File: packages/agent-forge/src/deploy/scope.ts
// Scope resolvers — each returns the `.claude/` root an artifact lands in, with
// the explicit input that scope requires. The artifact is identical across
// scopes; scope is the accident (`scope-grant`).
//
//   user     home resolver    -> <home>/.claude    (home defaults to $HOME)
//   project  project resolver -> <project>/.claude  (project defaults to cwd)
//
// The bare-home guard: `--home` is the user's HOME dir, so a bare home
// self-corrects — `.claude` appended, LOUDLY — while a path already ending in
// `.claude` is used verbatim (back-compat). This mirrors the remote placer's
// `resolveClaudeDir`. The footgun it fixes: before, LOCAL took an explicit
// `--home` verbatim, so `--home /Users/lex` littered `/Users/lex/{agents,skills}`
// beside the real `~/.claude`.

import { homedir } from 'node:os';
import { basename, resolve as resolvePath } from 'node:path';

export interface ScopeNote {
  message: string;
}

export interface ScopeResult {
  claudeDir: string;
  // A loud NOTE the caller prints to stderr when the bare-home guard fired.
  note: ScopeNote | null;
}

/** Expand a leading `~` to the user's home dir (POSIX tilde convenience). */
function expanduser(p: string): string {
  if (p === '~') {
    return homedir();
  }
  if (p.startsWith('~/')) {
    return resolvePath(homedir(), p.slice(2));
  }
  return p;
}

/** User scope claude dir: <home>/.claude. No `--home` → $HOME/.claude. A bare
 *  home self-corrects (`.claude` appended, with a loud NOTE); a path already
 *  ending in `.claude` is used verbatim. */
export function userScope(home?: string | null): ScopeResult {
  if (!home) {
    return { claudeDir: resolvePath(homedir(), '.claude'), note: null };
  }
  const p = resolvePath(expanduser(home));
  if (basename(p) === '.claude') {
    return { claudeDir: p, note: null };
  }
  const claudeDir = resolvePath(p, '.claude');
  return {
    claudeDir,
    note: {
      message: `  NOTE --home '${home}' is a home dir -> deploying to ${claudeDir}`,
    },
  };
}

/** Project scope claude dir: <project>/.claude (project defaults to cwd). */
export function projectScope(project?: string | null): ScopeResult {
  const root = project ? resolvePath(expanduser(project)) : process.cwd();
  return { claudeDir: resolvePath(root, '.claude'), note: null };
}
